components/sensor/ina219/ina219.go

```diff
@@ -0,0 +1,236 @@
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Warnf("using i2c address : %d", defaultI2Caddr)
+	}
+
+	s := &ina219{
+		name:    name,
+		logger:  logger,
+		bus:     i2cbus,
+		busName: attr.I2CBus,
+		addr:    byte(addr),
+	}
+
+	err = s.calibrate(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	busName    string
+	addr       byte
+	name       string
+	currentLSB int64
+	powerLSB   int64
+	cal        int64
+}
+
+type PowerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate(ctx context.Context) error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
+	// Calibration Register = 0.04096 / (current LSB * Shunt Resistance)
+	// Where lsb is in Amps and resistance is in ohms.
+	// Calibration register is 16 bits.
+	cal := calibratescale / (int64(d.currentLSB) * int64(senseResistor))
+	if cal >= (1 << 16) {
+		return fmt.Errorf("ina219 calibrate: calibration register value invalid %d", cal)
+	}
+	d.cal = cal
+
+	return nil
+}
+
+// Readings returns a list containing three items (voltage, current, and power).
+func (d *ina219) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	handle, err := d.bus.OpenHandle(d.addr)
+	if err != nil {
+		d.logger.Errorf("can't open ina219 i2c %s", err)
+		return nil, err
+	}
+
+	// calibrate sets the scaling factor of the current and power registers for the maximum resolution
+	buf := new(bytes.Buffer)
+	err = binary.Write(buf, binary.BigEndian, uint16(d.cal))
+	if err != nil {
+		return nil, err
+	}
+	err = handle.WriteBlockData(ctx, calibrationRegister, buf.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	buf = new(bytes.Buffer)
+	err = binary.Write(buf, binary.BigEndian, uint16(0x1FFF))
+	if err != nil {
+		return nil, err
+	}
+	err = handle.WriteBlockData(ctx, configRegister, buf.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	var pm PowerMonitor
+
+	// get shunt voltage - currently we are not returning - is it useful?
+	shunt, err := handle.ReadBlockData(ctx, shuntVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Least significant bit is 10µV.
+	pm.Shunt = int64(binary.BigEndian.Uint16(shunt)) * 10 * 1000
+
+	bus, err := handle.ReadBlockData(ctx, busVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if bit zero is set, if set the ADC has overflowed.
+	if binary.BigEndian.Uint16(bus)&1 > 0 {
+		return nil, fmt.Errorf("bus voltage register overflow")
```

Comment:
I'd put the word "ina219" in the error string somewhere. Otherwise, my first instinct is that something's wrong with the I2C bus.